src/utils/sources-tree/getDirectories.js

```diff
@@ -5,52 +5,46 @@
 // @flow
 
 import { createParentMap } from "./utils";
-import { getURL } from "./getURL";
 import type { TreeNode, TreeDirectory } from "./types";
 import type { Source } from "../../types";
 
-function findSource(sourceTree: TreeDirectory, sourceUrl: string): TreeNode {
-  let returnTarget = null;
-  function _traverse(subtree: TreeNode) {
-    if (subtree.type === "directory") {
-      for (const child of subtree.contents) {
-        _traverse(child);
-      }
-    } else if (!returnTarget) {
-      if (subtree.path.replace(/http(s)?:\//, "") == sourceUrl) {
-        returnTarget = subtree;
-        return;
-      }
+function _traverse(subtree: TreeNode, source: Source) {
+  if (subtree.type === "source") {
+    if (subtree.contents.id === source.id) {
+      return subtree;
     }
-  }
-
-  sourceTree.contents.forEach(node => _traverse(node));
 
-  if (!returnTarget) {
-    return sourceTree;
+    return null;
   }
 
-  return returnTarget;
+  const matches = subtree.contents.map(child => _traverse(child, source));
+  return matches && matches.filter(Boolean)[0];
 }
 
-export function getDirectories(source: Source, sourceTree: TreeDirectory) {
-  const url = getURL(source);
-  const fullUrl = `${url.group}${url.path}`;
-  const parentMap = createParentMap(sourceTree);
+function findSourceItem(sourceTree: TreeDirectory, source: Source): ?TreeNode {
+  return _traverse(sourceTree, source);
+}
 
-  const subtreeSource = findSource(sourceTree, fullUrl);
-  if (!subtreeSource) {
+function getAncestors(sourceTree: TreeDirectory, item: ?TreeNode) {
+  if (!item) {
     return [];
   }
 
-  let node = subtreeSource;
+  const parentMap = createParentMap(sourceTree);
   const directories = [];
-  directories.push(subtreeSource);
+
+  directories.push(item);
   while (true) {
-    node = parentMap.get(node);
-    if (!node) {
+    item = parentMap.get(item);
+    if (!item) {
       return directories;
     }
-    directories.push(node);
+    directories.push(item);
   }
 }
+
+export function getDirectories(source: Source, sourceTree: TreeDirectory) {
+  const item = findSourceItem(sourceTree, source);
+  const ancestors = getAncestors(sourceTree, item);
+  return ancestors && ancestors.length ? ancestors : [sourceTree];
```

Comment:
This change, ensuring the `[sourceTree]` is returned if no ancestors is found, keeps the functionality.

Reply:
Could this be `ancestors ? ancestors : [sourceTree];` because flow tells us that getAncestors is null or an array?